Bug: `sort-enums` breaks members referencing earlier members: "reference members declared after it,"

### Describe the bug

I'm seeing cases where `perfectionist/sort-enums` is asking to reorder an enum to be in alphabetical/natural order, even though an enum member has to go last to reference earlier members.

### Code example

```ts
enum ArgumentType {
  First = 1 << 0,
  RegExp = 1 << 1,
  Both = First | RegExp,
  //  ~~~~~~~~~~~~~~~~~ 
  // Expected "Both" to come before "RegExp". (eslint)
}
```

After fixing:

```ts
enum ArgumentType {
  Both = First | RegExp,
  First = 1 << 0,
  RegExp = 1 << 1,
}
```

```plaintext
$ npm run tsc

> repros@1.0.0 tsc
> tsc index.ts

index.ts:2:10 - error TS2651: A member initializer in a enum declaration cannot reference members declared after it, including members defined in other enums.

2   Both = First | RegExp,
           ~~~~~

index.ts:2:18 - error TS2651: A member initializer in a enum declaration cannot reference members declared after it, including members defined in other enums.

2   Both = First | RegExp,
                   ~~~~~~


Found 2 errors in the same file, starting at: index.ts:2
```

### ESLint version

9.0.0

### ESLint Plugin Perfectionist version

3.2.0

### Additional comments

Isolated repro: https://github.com/JoshuaKGoldberg/repros/tree/perfectionist-order-enums-self-referencing

### Validations

- [X] Read the [docs](https://perfectionist.dev).
- [X] Check that there isn't [already an issue](https://github.com/azat-io/eslint-plugin-perfectionist/issues) that reports the same bug to avoid creating a duplicate.

--- a/rules/sort-enums.ts
+++ b/rules/sort-enums.ts
@@ -159,23 +159,15 @@ export default createEslintRule<Options, MESSAGE_ID>({
             }
 
             if ('left' in nodeValue) {
-              traverseNode(nodeValue.left)
+              checkNode(nodeValue.left)
             }
 
             if ('right' in nodeValue) {
-              traverseNode(nodeValue.right)
+              checkNode(nodeValue.right)
             }
           }
 
-          let traverseNode = (nodeValue: TSESTree.Node[] | TSESTree.Node) => {
-            if (Array.isArray(nodeValue)) {
-              nodeValue.forEach(traverseNode)
-            } else {
-              checkNode(nodeValue)
-            }
-          }
-
-          traverseNode(expression)
+          checkNode(expression)
           return dependencies
         }
 

--- a/rules/sort-enums.ts
+++ b/rules/sort-enums.ts
@@ -128,6 +128,57 @@ export default createEslintRule<Options, MESSAGE_ID>({
         let sourceCode = getSourceCode(context)
         let partitionComment = options.partitionByComment
 
+        let extractDependencies = (
+          expression: TSESTree.Expression,
+          enumName: string,
+        ): string[] => {
+          let dependencies: string[] = []
+
+          let checkNode = (nodeValue: TSESTree.Node) => {
+            if (
+              nodeValue.type === 'MemberExpression' &&
+              nodeValue.object.type === 'Identifier' &&
+              nodeValue.object.name === enumName &&
+              nodeValue.property.type === 'Identifier'
+            ) {
+              /**
+               * enum Enum {
+               *   A = 1,
+               *   B = Enum.A
+               * }
+               */
+              dependencies.push(nodeValue.property.name)
+            } else if (nodeValue.type === 'Identifier') {
+              /**
+               * enum Enum {
+               *   A = 1,
+               *   B = A
+               * }
+               */
+              dependencies.push(nodeValue.name)
+            }
+
+            if ('left' in nodeValue) {
+              traverseNode(nodeValue.left)
+            }
+
+            if ('right' in nodeValue) {
+              traverseNode(nodeValue.right)
+            }
+          }
+
+          let traverseNode = (nodeValue: TSESTree.Node[] | TSESTree.Node) => {
+            if (Array.isArray(nodeValue)) {
+              nodeValue.forEach(traverseNode)
+            } else {
+              checkNode(nodeValue)
+            }
+          }
+
+          traverseNode(expression)
+          return dependencies
+        }
+
         let formattedMembers: SortingNode[][] = members.reduce(
           (accumulator: SortingNode[][], member) => {
             let comment = getCommentBefore(member, sourceCode)
@@ -145,9 +196,18 @@ export default createEslintRule<Options, MESSAGE_ID>({
                 ? `${member.id.value}`
                 : `${sourceCode.text.slice(...member.id.range)}`
 
-            let sortingNode = {
+            let dependencies
+            if (member.initializer) {
+              dependencies = extractDependencies(
+                member.initializer,
+                node.id.name,
+              )
+            }
+
+            let sortingNode: SortingNode = {
               size: rangeToDiff(member.range),
               node: member,
+              dependencies,
               name,
             }
             accumulator.at(-1)!.push(sortingNode)

--- a/test/sort-enums.test.ts
+++ b/test/sort-enums.test.ts
@@ -1703,5 +1703,79 @@ describe(ruleName, () => {
         ],
       },
     )
+
+    ruleTester.run(`${ruleName}: works with dependencies`, rule, {
+      valid: [
+        {
+          code: dedent`
+              enum Enum {
+                B = 'B',
+                A = B,
+              }
+            `,
+          options: [
+            {
+              type: 'alphabetical',
+            },
+          ],
+        },
+        {
+          code: dedent`
+              enum Enum {
+                B = 'B',
+                A = Enum.B,
+              }
+            `,
+          options: [
+            {
+              type: 'alphabetical',
+            },
+          ],
+        },
+        {
+          code: dedent`
+              enum Enum {
+                B = 0,
+                A = 1 | 2 | B | Enum.B,
+              }
+            `,
+          options: [
+            {
+              type: 'alphabetical',
+            },
+          ],
+        },
+      ],
+      invalid: [
+        {
+          code: dedent`
+              enum Enum {
+                B = 'B',
+                A = AnotherEnum.B,
+              }
+            `,
+          output: dedent`
+            enum Enum {
+              A = AnotherEnum.B,
+              B = 'B',
+            }
+          `,
+          options: [
+            {
+              type: 'alphabetical',
+            },
+          ],
+          errors: [
+            {
+              messageId: 'unexpectedEnumsOrder',
+              data: {
+                left: 'B',
+                right: 'A',
+              },
+            },
+          ],
+        },
+      ],
+    })
   })
 })

--- a/test/sort-enums.test.ts
+++ b/test/sort-enums.test.ts
@@ -1705,59 +1705,111 @@ describe(ruleName, () => {
     )
 
     ruleTester.run(`${ruleName}: works with dependencies`, rule, {
-      valid: [
-        {
-          code: dedent`
-              enum Enum {
-                B = 'B',
-                A = B,
-              }
-            `,
-          options: [
-            {
-              type: 'alphabetical',
-            },
-          ],
-        },
-        {
-          code: dedent`
-              enum Enum {
-                B = 'B',
-                A = Enum.B,
-              }
-            `,
-          options: [
-            {
-              type: 'alphabetical',
-            },
-          ],
-        },
-        {
-          code: dedent`
-              enum Enum {
-                B = 0,
-                A = 1 | 2 | B | Enum.B,
-              }
-            `,
-          options: [
-            {
-              type: 'alphabetical',
-            },
-          ],
-        },
-      ],
+      valid: [],
       invalid: [
         {
           code: dedent`
-              enum Enum {
-                B = 'B',
-                A = AnotherEnum.B,
-              }
-            `,
+            enum Enum {
+              C = 'C',
+              B = 0,
+              A = B,
+            }
+          `,
+          output: dedent`
+            enum Enum {
+              B = 0,
+              A = B,
+              C = 'C',
+            }
+          `,
+          options: [
+            {
+              type: 'alphabetical',
+            },
+          ],
+          errors: [
+            {
+              messageId: 'unexpectedEnumsOrder',
+              data: {
+                left: 'C',
+                right: 'B',
+              },
+            },
+          ],
+        },
+        {
+          code: dedent`
+            enum Enum {
+              C = 'C',
+              B = 0,
+              A = Enum.B,
+            }
+          `,
+          output: dedent`
+            enum Enum {
+              B = 0,
+              A = Enum.B,
+              C = 'C',
+            }
+          `,
+          options: [
+            {
+              type: 'alphabetical',
+            },
+          ],
+          errors: [
+            {
+              messageId: 'unexpectedEnumsOrder',
+              data: {
+                left: 'C',
+                right: 'B',
+              },
+            },
+          ],
+        },
+        {
+          code: dedent`
+            enum Enum {
+              C = 3,
+              B = 0,
+              A = 1 | 2 | B | Enum.B,
+            }
+          `,
+          output: dedent`
+            enum Enum {
+              B = 0,
+              A = 1 | 2 | B | Enum.B,
+              C = 3,
+            }
+          `,
+          options: [
+            {
+              type: 'alphabetical',
+            },
+          ],
+          errors: [
+            {
+              messageId: 'unexpectedEnumsOrder',
+              data: {
+                left: 'C',
+                right: 'B',
+              },
+            },
+          ],
+        },
+        {
+          code: dedent`
+            enum Enum {
+              B = 'B',
+              A = AnotherEnum.B,
+              C = 'C',
+            }
+          `,
           output: dedent`
             enum Enum {
               A = AnotherEnum.B,
               B = 'B',
+              C = 'C',
             }
           `,
           options: [

